spatialmath/capsule.go

```diff
@@ -0,0 +1,346 @@
+		if dist < lowDist {
+			lowDist = dist
+		}
+	}
+	return lowDist
+}
+
+func capsuleVsTriangleDistance(c *capsule, other *triangle) float64 {
+	capPt, triPt := closestPointsSegmentTriangle(c.segA, c.segB, other)
+	return capPt.Sub(triPt).Norm() - c.radius
+}
+
+func capsuleVsBoxCollision(c *capsule, other *box) bool {
+	// TODO(pl): Large amounts of capsule collision code were adopted from `brax`
+	// https://github.com/google/brax/blob/7eaa16b4bf446b117b538dbe9c9401f97cf4afa2/brax/physics/colliders.py
+	// https://github.com/google/brax/blob/7eaa16b4bf446b117b538dbe9c9401f97cf4afa2/brax/physics/geometry.py
+	// Brax converts boxes to meshes composed of 12 triangles and does collision detection on those.
+	// We support doing so but it appears that separating axis test is superior so that is what is used here.
+	// The two methods should be fully characterized.
+	return capsuleBoxSeparatingAxis(c, other)
+}
+
+// capsuleInCapsule returns a bool describing if the inner capsule is fully encompassed by the outer capsule.
+func capsuleInCapsule(inner, outer *capsule) bool {
+	return capsuleVsPointDistance(outer, inner.segA) < -inner.radius &&
+		capsuleVsPointDistance(outer, inner.segB) < -inner.radius
+}
+
+// capsuleInBox returns a bool describing if the given capsule is fully encompassed by the given box.
+func capsuleInBox(c *capsule, b *box) bool {
+	return pointVsBoxDistance(c.segA, b) <= -c.radius && pointVsBoxDistance(c.segB, b) <= -c.radius
+}
+
+// capsuleInSphere returns a bool describing if the given capsule is fully encompassed by the given sphere.
+func capsuleInSphere(c *capsule, s *sphere) bool {
+	return c.segA.Sub(s.pose.Point()).Norm()+c.radius <= s.radius && c.segB.Sub(s.pose.Point()).Norm()+c.radius <= s.radius
+}
+
+// capsuleBoxSeparatingAxis returns immediately as soon as any result is found indicating that the two objects are not in collision.
+func capsuleBoxSeparatingAxis(a *capsule, b *box) bool {
+	capCenter := a.segA.Add(a.segB).Mul(0.5)
+	centerDist := b.pose.Point().Sub(capCenter)
+
+	// check if there is a distance between bounding spheres to potentially exit early
+	if centerDist.Norm()-((a.length/2)+b.boundingSphereR) > CollisionBuffer {
+		return false
+	}
+	rmA := a.pose.Orientation().RotationMatrix()
+	rmB := b.pose.Orientation().RotationMatrix()
+
+	segLen := (a.length - 2*a.radius) / 2
+	capVec := rmA.Row(2).Mul(segLen)
+
+	for i := 0; i < 3; i++ {
+		if separatingAxisTest1D(centerDist, rmA.Row(i), capVec, b.halfSize, rmB) > CollisionBuffer+a.radius {
+			return false
+		}
+		if separatingAxisTest1D(centerDist, rmB.Row(i), capVec, b.halfSize, rmB) > CollisionBuffer+a.radius {
+			return false
+		}
+		for j := 0; j < 3; j++ {
+			crossProductPlane := rmA.Row(i).Cross(rmB.Row(j))
+
+			// if edges are parallel, this check is already accounted for by one of the face projections, so skip this case
+			if !utils.Float64AlmostEqual(crossProductPlane.Norm(), 0, floatEpsilon) {
```

Comment:
I've seen this line 3 places by now, would it maybe be better to move this into the separating axis test? or is it purposefully not in there because it only gets hit on these crossProductPlanes?

Reply:
Not only the cross product reason, but also we avoid an entire function call.